### PR TITLE
Wqcoprocess

### PR DESCRIPTION
**Pull request by David-Simonetti-ND**
Initial PR for issue #2825

### CODE REVIEW

**Comment by David-Simonetti-ND** [comment]
I am unhappy with how the end_coprocess code is right now because it is very much a copy paste of the code work_queue_factory uses. would there be some way to unify these functions?

**Comment by btovar** [comment]
@David-Simonetti-ND, yes, that's a good idea.  Could you add such a function to `dttools/src/process.*` with a name such as `process_kill_waitpid(pid, timeout)` ?  It should work in the same way as `process_waitpid`.

File: dttools/src/process.c [hunk 1]
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
...
+                sigemptyset(&new_action.sa_mask);
+                new_action.sa_flags = 0;
+                sigaction(SIGALRM, &new_action, &old_action);
+        }
+	int num_attempts = 0; // keep track of how many time we attempted to signal process pid
+	int current_signal = SIGTERM; // start by sending SIGTERM, move on to SIGKILL
+	int status; 
+	int ret_val = 1; // return value from this function: 0 means process exited cleanly, 1 means there was an error
+	while (kill(pid, current_signal) == 0) // send signal to process
+	{
```
Comment (btovar) [inline dttools/src/process.c:120]: to match the style of the other functions, move this brace to previous line

File: dttools/src/process.c [hunk 2]
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
...
+		if (timeout != 0) {alarm(timeout);} // and set an alarm for timeout seconds
```
Comment (btovar) [inline dttools/src/process.c:121]: expand to three lines ( if {..., alarm,   } )

File: dttools/src/process.c [hunk 3]
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
...
+		num_attempts++;
+		if (num_attempts == max_attempts) // change to sending SIGKILL after max_attempts tries
```
Comment (btovar) [inline dttools/src/process.c:123]: move brace to previous line, same for all the rest of the function

Comment (btovar) [reply dttools/src/process.c:123]: Also, `if(num_attempts >= max_attempts)`

File: dttools/src/process.c [hunk 4]
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
 	return 0;
 }
 
+int process_kill_waitpid(pid_t pid, int timeout, int max_attempts)
+{
+	int flags = 0;
+        struct sigaction new_action, old_action;
```
Comment (btovar) [inline dttools/src/process.c:104]: Is the file function indented uniformly? The rest of the file uses only tabs.

**Review by btovar — CHANGES_REQUESTED** [review]

File: dttools/src/process.c [hunk 5]
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
...
+			if (errno == EINTR) // this is the case that the alarm woke us up so we try again to signal
+			{
+				continue;
+			}
+			else // if there is any other error, it is fatal and we should just exit
+			{
+				break;
+			}
+		}
+		else if (return_pid == 0) // if the return value is 0, it means we set the timeout to zero and nothing changed, so we try to signal and wait for the process again
```
Comment (btovar) [inline dttools/src/process.c:143]: Doesn't this trigger a busy-wait? I.e.. If the timeout is zero, and waitpid is no hang, then if the processes has not changed, this loop will be executed over and over again with no pause in between.

Comment (btovar) [reply dttools/src/process.c:143]: I think it should be ok to always require `process_kill_waitpid` to have a positive timeout and use waitpid without the no hang flag. Also, I'm not sure about the re-sending of signals until max_attempts. Shouldn't we just send the signal once and then do the sigkill?  What worries me is that a processes may modify its cleanup when receiving the same signal in succession (e.g. first signal triggers cleanup, second or more terminates immediately).

**Review by btovar — CHANGES_REQUESTED** [review]

**Comment by David-Simonetti-ND** [comment]
I reworked process_kill_waitpid to use process_waitpid because there may be cases where a given child has already return and is in the completed process list. It now only sends SIGTERM once, waits for timeout seconds, and then sends SIGKILL and attempts to wait for another timeout seconds. At the end it just gives up.
I also added functionality in the work_queue_worker for checking if the coprocess specified in the command line has stopped running and restarting it if this is the case.

**Comment by btovar** [comment]
Looks good. Ready to merge?

**Review by btovar — APPROVED** [review]

**Pull request merged**